File: android/src/main/kotlin/io/agora/agorartm/RTMCallManager.kt
package io.agora.agorartm

import android.os.Handler
import io.agora.rtm.LocalInvitation
import io.agora.rtm.RemoteInvitation
import io.agora.rtm.RtmCallEventListener
import io.agora.rtm.RtmCallManager
import io.agora.rtm.RtmClient
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel

class RTMCallManager(
    client: RtmClient,
    clientIndex: Long,
    messenger: BinaryMessenger,
    private val handler: Handler
) : RtmCallEventListener, EventChannel.StreamHandler {
    private val eventChannel: EventChannel =
        EventChannel(messenger, "io.agora.rtm.client${clientIndex}.call_manager")
    private var eventSink: EventChannel.EventSink? = null

    val manager: RtmCallManager
    val remoteInvitations: MutableMap<Int?, RemoteInvitation> = HashMap()
    val localInvitations: MutableMap<Int?, LocalInvitation> = HashMap()

    init {
        this.eventChannel.setStreamHandler(this)

        manager = client.rtmCallManager
        manager.setEventListener(this)
    }

    private fun sendEvent(eventName: String, params: HashMap<String, Any?>) {
        handler.post {
            this.eventSink?.success(
                hashMapOf(
                    "event" to eventName,
                    "data" to params,
                )
            )
        }
    }

    override fun onListen(params: Any?, eventSink: EventChannel.EventSink) {
        this.eventSink = eventSink
    }

    override fun onCancel(params: Any?) {
        this.eventSink = null
    }

    override
    fun onLocalInvitationReceivedByPeer(localInvitation: LocalInvitation) {
        localInvitations[localInvitation.hashCode()] = localInvitation
        sendEvent(
            "onLocalInvitationReceivedByPeer", hashMapOf(
                "localInvitation" to localInvitation.toJson()
            )
        )
    }

    override
    fun onLocalInvitationAccepted(localInvitation: LocalInvitation, response: String) {
        localInvitations.remove(localInvitation.hashCode())
        sendEvent(
            "onLocalInvitationAccepted", hashMapOf(
                "localInvitation" to localInvitation.toJson(),
                "response" to response,
            )
        )
    }

    override
    fun onLocalInvitationRefused(localInvitation: LocalInvitation, response: String) {
        localInvitations.remove(localInvitation.hashCode())
        sendEvent(
            "onLocalInvitationRefused", hashMapOf(
                "localInvitation" to localInvitation.toJson(),
                "response" to response,
            )
        )
    }

    override
    fun onLocalInvitationCanceled(localInvitation: LocalInvitation) {
        localInvitations.remove(localInvitation.hashCode())
        sendEvent(
            "onLocalInvitationCanceled", hashMapOf(
                "localInvitation" to localInvitation.toJson()
            )
        )
    }

    override
    fun onLocalInvitationFailure(localInvitation: LocalInvitation, errorCode: Int) {
        sendEvent(
            "onLocalInvitationFailure", hashMapOf(
                "localInvitation" to localInvitation.toJson(),
                "errorCode" to errorCode,
            )
        )
    }

    override
    fun onRemoteInvitationReceived(remoteInvitation: RemoteInvitation) {
        remoteInvitations[remoteInvitation.hashCode()] = remoteInvitation
        sendEvent(
            "onRemoteInvitationReceived", hashMapOf(
                "remoteInvitation" to remoteInvitation.toJson()
            )
        )
    }

    override
    fun onRemoteInvitationAccepted(remoteInvitation: RemoteInvitation) {
        remoteInvitations.remove(remoteInvitation.hashCode())
        sendEvent(
            "onRemoteInvitationAccepted", hashMapOf(
                "remoteInvitation" to remoteInvitation.toJson()
            )
        )
    }

    override
    fun onRemoteInvitationRefused(remoteInvitation: RemoteInvitation) {
        remoteInvitations.remove(remoteInvitation.hashCode())
        sendEvent(
            "onRemoteInvitationRefused", hashMapOf(
                "remoteInvitation" to remoteInvitation.toJson()
            )
        )
    }

    override
    fun onRemoteInvitationCanceled(remoteInvitation: RemoteInvitation) {
        remoteInvitations.remove(remoteInvitation.hashCode())
        sendEvent(
            "onRemoteInvitationCanceled", hashMapOf(
                "remoteInvitation" to remoteInvitation.toJson()
            )
        )
    }

    override
    fun onRemoteInvitationFailure(remoteInvitation: RemoteInvitation, errorCode: Int) {
        sendEvent(
            "onRemoteInvitationFailure", hashMapOf(
                "remoteInvitation" to remoteInvitation.toJson(),
                "errorCode" to errorCode,
            )
        )
    }
}

File: ios/Classes/RTMCallManager.swift
//
//  RTMCallManager.swift
//  agora_rtm
//
//  Created by LXH on 2023/5/31.
//

import Foundation
import AgoraRtmKit
import Flutter

class RTMCallManager: NSObject, AgoraRtmCallDelegate, FlutterStreamHandler {
    var eventChannel: FlutterEventChannel?
    var eventSink: FlutterEventSink?
    
    var client: AgoraRtmKit?
    var remoteInvitations: [Int: AgoraRtmRemoteInvitation] = [:]
    var localInvitations: [Int: AgoraRtmLocalInvitation] = [:]
    var manager: AgoraRtmCallKit? {
        get {
            return client?.getRtmCall()
        }
    }
    
    init(_ client: AgoraRtmKit, _ clientIndex: Int, _ messenger: FlutterBinaryMessenger) {
        eventChannel = FlutterEventChannel(
            name: "io.agora.rtm.client\(clientIndex).call_manager",
            binaryMessenger: messenger
        )
        self.client = client
        super.init()
        eventChannel?.setStreamHandler(self)
        self.manager?.callDelegate = self
    }
    
    deinit {
        remoteInvitations.removeAll()
        localInvitations.removeAll()
        eventChannel = nil
        client = nil
    }
    
    private func sendEvent(eventName: String, params: [String: Any?]) {
        let event: [String: Any?] = [
            "event": eventName,
            "data": params,
        ]
        
        eventSink?(event)
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
    
    func rtmCallKit(_ callKit: AgoraRtmCallKit, localInvitationReceivedByPeer localInvitation: AgoraRtmLocalInvitation) {
        localInvitations[localInvitation.hash] = localInvitation
        sendEvent(eventName: "onLocalInvitationReceivedByPeer", params: [
            "localInvitation": localInvitation.toJson(),
        ])
    }
    
    func rtmCallKit(_ callKit: AgoraRtmCallKit, localInvitationAccepted localInvitation: AgoraRtmLocalInvitation, withResponse response: String?) {
        localInvitations.removeValue(forKey: localInvitation.hash)
        sendEvent(eventName: "onLocalInvitationAccepted", params: [
            "localInvitation": localInvitation.toJson(),
            "response": response,
        ])
    }
    
    func rtmCallKit(_ callKit: AgoraRtmCallKit, localInvitationRefused localInvitation: AgoraRtmLocalInvitation, withResponse response: String?) {
        localInvitations.removeValue(forKey: localInvitation.hash)
        sendEvent(eventName: "onLocalInvitationRefused", params: [
            "localInvitation": localInvitation.toJson(),
            "response": response,
        ])
    }
    
    func rtmCallKit(_ callKit: AgoraRtmCallKit, localInvitationCanceled localInvitation: AgoraRtmLocalInvitation) {
        localInvitations.removeValue(forKey: localInvitation.hash)
        sendEvent(eventName: "onLocalInvitationCanceled", params: [
            "localInvitation": localInvitation.toJson(),
        ])
    }
    
    func rtmCallKit(_ callKit: AgoraRtmCallKit, localInvitationFailure localInvitation: AgoraRtmLocalInvitation, errorCode: AgoraRtmLocalInvitationErrorCode) {
        sendEvent(eventName: "onLocalInvitationFailure", params: [
            "localInvitation": localInvitation.toJson(),
            "errorCode": errorCode.rawValue,
        ])
    }
    
    func rtmCallKit(_ callKit: AgoraRtmCallKit, remoteInvitationReceived remoteInvitation: AgoraRtmRemoteInvitation) {
        remoteInvitations[remoteInvitation.hash] = remoteInvitation
        sendEvent(eventName: "onRemoteInvitationReceived", params: [
            "remoteInvitation": remoteInvitation.toJson(),
        ])
    }
    
    func rtmCallKit(_ callKit: AgoraRtmCallKit, remoteInvitationAccepted remoteInvitation: AgoraRtmRemoteInvitation) {
        remoteInvitations.removeValue(forKey: remoteInvitation.hash)
        sendEvent(eventName: "onRemoteInvitationAccepted", params: [
            "remoteInvitation": remoteInvitation.toJson(),
        ])
    }
    
    func rtmCallKit(_ callKit: AgoraRtmCallKit, remoteInvitationRefused remoteInvitation: AgoraRtmRemoteInvitation) {
        remoteInvitations.removeValue(forKey: remoteInvitation.hash)
        sendEvent(eventName: "onRemoteInvitationRefused", params: [
            "remoteInvitation": remoteInvitation.toJson(),
        ])
    }
    
    func rtmCallKit(_ callKit: AgoraRtmCallKit, remoteInvitationCanceled remoteInvitation: AgoraRtmRemoteInvitation) {
        remoteInvitations.removeValue(forKey: remoteInvitation.hash)
        sendEvent(eventName: "onRemoteInvitationCanceled", params: [
            "remoteInvitation": remoteInvitation.toJson(),
        ])
    }
    
    func rtmCallKit(_ callKit: AgoraRtmCallKit, remoteInvitationFailure remoteInvitation: AgoraRtmRemoteInvitation, errorCode: AgoraRtmRemoteInvitationErrorCode) {
        sendEvent(eventName: "onRemoteInvitationFailure", params: [
            "remoteInvitation": remoteInvitation.toJson(),
            "errorCode": errorCode.rawValue,
        ])
    }
}
